File: server/app/schemas/treasure_chest.py
from datetime import datetime
from enum import Enum

from pydantic import BaseModel


class ChestState(str, Enum):
    NO_CHEST = "no_chest"
    LOCKED = "locked"
    OPENABLE = "openable"
    OPENED = "opened"


class TreasureChestResponse(BaseModel):
    state: ChestState
    armed_at: datetime | None
    openable_at: datetime | None
    opened_at: datetime | None
    reward_gems: int
    remaining_seconds: int | None


class OpenChestResponse(BaseModel):
    reward_gems: int
    balance: int
    opened_at: datetime

File: app/lib/features/gems/models/treasure_chest.dart
// ignore_for_file: invalid_annotation_target

import 'package:freezed_annotation/freezed_annotation.dart';
import 'chest_state.dart';

part 'treasure_chest.freezed.dart';
part 'treasure_chest.g.dart';

@freezed
class TreasureChest with _$TreasureChest {
  const factory TreasureChest({
    required ChestState state,
    @JsonKey(name: 'armed_at') DateTime? armedAt,
    @JsonKey(name: 'openable_at') DateTime? openableAt,
    @JsonKey(name: 'opened_at') DateTime? openedAt,
    @JsonKey(name: 'reward_gems') required int rewardGems,
    @JsonKey(name: 'remaining_seconds') int? remainingSeconds,
  }) = _TreasureChest;

  factory TreasureChest.fromJson(Map<String, dynamic> json) =>
      _$TreasureChestFromJson(json);
}
